File: examples/Inheritance.java
// ----------------
// Inheritance.java
// ----------------

import java.util.ArrayList;

class Shape {
    private int _x;
    private int _y;

    public Shape (int x, int y) {
        _x = x;
        _y = y;}

    public double area () {
        return 0;}

    public final void move (int x, int y) {
        _x = x;
        _y = y;}

    public String toString () {
        return _x + ", " + _y;}}

final class Circle extends Shape {
    private int _r;

    public Circle (int x, int y, int r) {
        super(x, y);
        _r = r;}

    public double area () {
        return 3.14 * _r * _r;}

    public int radius () {
        return _r;}

    public String toString () {
        return "(" + super.toString() + ", " + _r + ")";}}

final class Inheritance {
    public static void main (String[] args) {
        System.out.println("Inheritance.java");

        {
        final Shape x = new Shape(2, 3);
        assert x.area()     == 0;
        x.move(5, 6);
//      assert x.radius()   == 0;           // error: cannot find symbol
        assert x.toString().equals("5, 6");
        }

        {
        final Circle x = new Circle(2, 3, 4);
        assert x.area()   == 3.14 * 4 * 4;
        x.move(5, 6);
        assert x.radius() == 4;
        assert x.toString().equals("(5, 6, 4)");
        }

        {
        final Shape x = new Circle(2, 3, 4);
        assert x.area()              == 3.14 * 4 * 4;
        x.move(5, 6);
//      assert x.radius()            == 0;            // error: cannot find symbol
        assert ((Circle) x).radius() == 4;
        assert x.toString().equals("(5, 6, 4)");
        }

        {
//      final Circle[] a = {new Shape(2, 3), new Circle(2, 3, 4)}; // error: incompatible types
        final Shape[]  a = {new Shape(2, 3), new Circle(2, 3, 4)};
        assert a[0].area() == 0;
        assert a[1].area() == 3.14 * 4 * 4;
        }

        {
        final Shape[]          a = new Circle[3];
//      final ArrayList<Shape> x = new ArrayList<Circle>(10); // error: incompatible types
        }

        System.out.println("Done.");}}

File: examples/Inheritance.py
#!/usr/bin/env python

# --------------
# Inheritance.py
# --------------

print "Inheritance.py"

class Shape (object) :
    def __init__ (self, x, y) :
        super(Shape, self).__init__()
        self.x = x
        self.y = y

    def area (self) :
        return 0

    def move (self, x, y) :
        self.x = x
        self.y = y

    def __str__ (self) :
        return str(self.x) + ", " + str(self.y)

class Circle (Shape) :
    def __init__ (self, x, y, r) :
        super(Circle, self).__init__(x, y) # doesn't mention Shape
#       Shape.__init__(self, x, y)
        self.r = r

    def area (self) :
        return 3.14 * self.r * self.r

    def radius (self) :
        return self.r

    def __str__ (self) :
        return "(" + super(Circle, self).__str__() + ", " + str(self.r) + ")"
#       return "(" + Shape.__str__(self)           + ", " + str(self.r) + ")"

class PseudoCircle (Shape) :
    def __init__ (self, r) :
        self.r = r

x = Shape(2, 3)
assert x.area()          == 0
x.move(5, 6)
#assert x.radius()       == 0      # AttributeError: 'Shape' object has no attribute 'radius'
#assert Circle.radius(x) == 0      # TypeError: unbound method radius() must be called with Circle instance as first argument (got Shape instance instead)
assert str(x)            == "5, 6"

x = Circle(2, 3, 4)
assert x.area()     == 3.14 * 4 * 4
x.move(5, 6)
assert x.radius()   == 4
assert str(x)       == "(5, 6, 4)"

t = (Shape(2, 3), Circle(2, 3, 4))
assert t[0].area() == 0
assert t[1].area() == 3.14 * 4 * 4

x = PseudoCircle(4)
#assert str(x) == "5, 6" # AttributeError: 'PseudoCircle' object has no attribute 'x'
x.move(5, 6)
assert str(x)  == "5, 6"

print "Done."
